Avoid dynamic_cast in ghost/artificial cell queries

kronbichler opened the pull request:
This replaces the `dynamic_cast` in the `CellAccessor::is_locally_owned` and friends functions by a query to the subdomain of the triangulation as discussed in #8801. While that is also a virtual function call, it is still much cheaper than finding out about the actual type of a class with virtual inheritance. As discussed in #8801, we do not intend to change the subdomain id of a serial triangulation at this point.

Closes #8801.

kronbichler commented:
I ran all tests with `mpi, grid, tria` in the name and they all passed. I will run the full testsuite now.

kronbichler commented:
All tests passing.

bangerth commented on include/deal.II/grid/tria_accessor.templates.h:
```diff
@@ -3566,17 +3566,10 @@ CellAccessor<dim, spacedim>::is_locally_owned() const
-  else
-    return (this->subdomain_id() == pt->locally_owned_subdomain());
+  return (this->tria->locally_owned_subdomain() ==
+            numbers::invalid_subdomain_id ||
```
Can I ask you to prefix all of these comparisons with a comment that says that serial triangulations report `numbers::invalid_subdomain_id` for `tria->locally_owned_subdomain()`, and that consequently the first part of the condition indicates that the triangulation is serial where all cells are locally owned?

Same in the other places.

bangerth commented:
This looks good in principle, but I think a comment in all of the places of concern would help understand what the condition really tests.

kronbichler commented:
I added the suggested comments. Furthermore, I also removed the inclusion to the now obsolete header file `distributed/tria_base.h` in `tria_accessor.templates.h`. As a consequence, I had to add two includes elsewhere to make sure `parallel::TriangulationBase` (in `dof_handler.h`) and `ArrayView` (in `mapping.h`, that was previously transitively included via `tria_accessor.templates.h` -> `tria.h` -> `distributed/tria_base.h` -> `base/mpi.h` -> `base/array_view.h`). I'm re-running the test suite now because we might have depended on this include somewhere as well, given that we also did in the library.
